### DISCUSSION TITLE
Async updates in fieldArrays that may change their indexes during async cycle

### DISCUSSION THREAD

**Discussion by cbdeveloper** [react-hook-form/react-hook-form#8151] [Q&A]
I have a array field that I use to upload images.

This is the type of a single image input

```
type ImageField = {
  src: string,
  size: number,
  alt: string
}
```

In my use-case, the user can add/remove/move images around, basically adding/removing and re-ordering the array.

This should be taken care by these methods right?

https://react-hook-form.com/api/usefieldarray

![image](https://user-images.githubusercontent.com/43407798/162372328-fc3a79d4-f29f-4e32-8a75-62ec9066a09c.png)

But the issue is that some of the updates are `async`, because I'm uploading the images on the fly.

This is the flow:

- User selects an image
- Image is uploaded to storage
- Image field is updated with new `src` and `size`

I should use the `update` method to do this right?

But it seems that that method uses `index` to identify the element. And what if the indexes have changes during the `async` cycle of my update? Will that still work? Or will I be updating a different input?

![image](https://user-images.githubusercontent.com/43407798/162372573-e2fceba6-75e7-4218-9872-288284ac7d32.png)

Shouldn't the signature of this method be something like:

```
update 

(id : string, obj: object) => void
([keyName] : string, obj: object) => void // If I pass keyName to `useFieldArray`
```

**UPDATE**

CodeSandbox [link](https://codesandbox.io/s/nice-shadow-ufz1y6?file=/src/App.tsx)

https://user-images.githubusercontent.com/43407798/162685412-3ec23f9c-8dc9-462c-9beb-690283cb96ca.mov

**Comment by bluebill1049** [comment]
can you share a simple codesandbox?

**Reply by cbdeveloper** [reply]
Hey @bluebill1049 , thanks for your reply.

Here is a code CodeSandBox with the example.

https://codesandbox.io/s/nice-shadow-ufz1y6?file=/src/App.tsx

**Comment by Dragomir-Ivanov** [comment]
Then you find that index you need!

You will need a dedicated `id` field in your array items. You can pick just any function that can generate random numbers. You will not have millions of items in that array, so it doesn't need to be that strong.

You tell RHF that it needs to use another field name than `id`, for its internal workings:
```
  const { fields, append, remove, update } = useFieldArray({
    ......
    keyName: "_keyId",
  })

```

Then where you modify the item in the array, find the index you need:
```
 const idx = fields.findIndex((f) => f.id! === MyItem.id)
          if (idx !== -1) {
            update(idx, MyItem)
          }
```

The way I do with current RHF:

```
const handleClose = useCallback(
    (asset: Asset | undefined) => {
      if (asset) {
        if (asset.id) {
          const idx = fields.findIndex((f) => f.id! === asset.id)
          if (idx !== -1) {
            update(idx, asset)
          }
        } else {
          asset.id = generateId()
          append(asset)
        }
      }
    },
    [fields, append, update]
  )
```

**Reply by cbdeveloper** [reply]
Thanks for your reply. Sure we can create our own `updateById` functions, but I think this would make more sense if it would come out of the box from RHF. The idea is that indexes may change at any time, `ids` can't. 

If the `id` is no longer present (because the item might have been removed), I think the update should log the error to the console and that's it.

Maybe we could expose both `update` and `updateById`. We could also do the same with other methods (expose the `byId` variation). Does it make sense?
